src/nvim/syntax.c

```diff
@@ -6714,6 +6714,66 @@ int lookup_color(const int idx, const bool foreground, TriState *const boldp)
   return color;
 }
 
+void set_hl_group(int id, HlAttrs attrs, int link_id)
+{
+    int idx = id - 1;  // Index is ID minus one.
+
+    bool is_default = attrs.rgb_ae_attr & HL_DEFAULT;
+
+    // Return if "default" was used and the group already has settings
+    if (is_default && hl_has_settings(idx, true)) {
+      return;
+    }
+
+    HlGroup *g = &HL_TABLE()[idx];
+
+    if (link_id > 0) {
+      g->sg_cleared = false;
+      g->sg_link = link_id;
+      return;
+    }
+
+    g->sg_cleared = false;
+    g->sg_link = 0;
+    g->sg_gui = attrs.rgb_ae_attr;
+
+    g->sg_rgb_bg = attrs.rgb_bg_color;
+    g->sg_rgb_fg = attrs.rgb_fg_color;
+    g->sg_rgb_sp = attrs.rgb_sp_color;
+
+    char name[8];
+
+    struct {
+      char **name; RgbValue *val;
+    } cattrs[] = {
+      { &g->sg_rgb_bg_name, &g->sg_rgb_bg },
+      { &g->sg_rgb_fg_name, &g->sg_rgb_fg },
+      { &g->sg_rgb_sp_name, &g->sg_rgb_sp },
+      { NULL, 0 },
+    };
+
+    for (int j = 0; cattrs[j].name; j++) {
+      if (*cattrs[j].val != -1) {
+        xfree(*cattrs[j].name);
+        snprintf(name, sizeof(name), "#%06x", *cattrs[j].val);
```

Comment:
hmm we should preserve a named color here (otherwise it is not much meaning with having the `sg_rgb_bg_name` fields). If we changed the dict to a keyset we could just reach in and copy whatever the string fields was.

Reply:
Admittedly, I only did this as a low effort solution to get `:highlight [name]<cr>` to work correctly since its value is used as a condition in places, e.g: https://github.com/neovim/neovim/blob/1907a9481423ccf4433f8184bbfdc633589e43f5/src/nvim/syntax.c#L7604-L7606

But yeah, Ideally we should keep the name if we can.

Can you spoon-feed me a bit on how to change the dict to a keyset? Is there an example in the codebase I can work off?

Reply:
I ended up just adding a helper variable of type `HlAttrNames` to pass through the color names.

Reply:
Sure, I'll refactor the code later :]